{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4a7c8420-f229-4557-8aa3-0cb9bdd9cfae",
   "metadata": {},
   "source": [
    "## Finding Hidden Messages in DNA (Bioinformatics I)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f72e3fb5-1151-48b1-857c-2577361a5ee5",
   "metadata": {},
   "source": [
    "#### Week 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aca60f3-0482-4eea-882e-542efbacceef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The function PatternCount(text, pattern) “slides a window” down Text, checking whether each k-mer substring of Text matches Pattern.\n",
    "\n",
    "def PatternCount(text, pattern):  \n",
    "    count = 0  \n",
    "    for i in range(len(text)-len(pattern)+1):  \n",
    "        if text[i:i+len(pattern)] == pattern:  \n",
    "            count=count+1  \n",
    "    return count  \n",
    "print (PatternCount(text, pattern)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "128b3aa5-7844-4e05-b763-2f0d173abc55",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The function FrequencyMap(text, k) takes a string text and an integer k as input \n",
    "# and returns their frequency table as a map of string keys to integer values.\n",
    "\n",
    "def FrequencyMap(text, k):\n",
    "    dict_0 = {}\n",
    "    for i in range(len(text)-k+1):\n",
    "        if text[i:(i+k)] not in dict_0.keys():\n",
    "            dict_0[text[i:(i+k)]] = 1\n",
    "        else:\n",
    "            dict_0[text[i:(i+k)]] += 1\n",
    "    return(dict_0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2be6756-4de5-4c2d-9499-fc4abd05fcae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#finding the maximum value of FrequencyMap.\n",
    "\n",
    "MaxMap = max(dict_seq.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d5e5131-26b5-490d-83bb-2bce82049861",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The function BetterFrequentWords(text, k) finds all frequent k-mers in text.\n",
    "\n",
    "def BetterFrequentWords(text, k):   \n",
    "    frequentPatterns = []   \n",
    "    freqMap = FrequencyMap(text, k)   \n",
    "    MaxMap = max(dict_seq.values())   \n",
    "    for most_frequent_w in freqMap:   \n",
    "        if freq[most_frequent_w] == MaxMap:\n",
    "            frequentPatterns.append(most_frequent_w)\n",
    "    return(frequentPatterns)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48d75c07-3a50-46af-98c2-87ea536c5923",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function ReverseComplement(text) finds the reverse complement of a DNA string\n",
    "\n",
    "def ReverseComplement(text):\n",
    "    letters = { \"A\":\"T\", \"C\":\"G\", \"G\":\"C\", \"T\":\"A\"}\n",
    "    seq_rev = ''\n",
    "    for i in range(len(text)):\n",
    "        seq_rev += letters[text[len(text) - i - 1]]\n",
    "    return seq_rev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a643bcc-6243-4f3b-8787-d212eba49c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function PatternPosition(genome, pattern) finds all occurrences of a pattern in a string (GENOME)\n",
    "\n",
    "def PatternPosition(GENOME, pattern):\n",
    "    position = []\n",
    "    for i in range(0, len(GENOME)- len(pattern)):\n",
    "        if GENOME[i: i + len(pattern)] == pattern:\n",
    "            position.append(i)\n",
    "    return position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d22e543d-8ca6-42a4-9acd-0cb42bac0968",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function FindClumps(GENOME, K, L, T) finds patterns forming clumps ((L, T)-clump inside a (longer) string\n",
    "# GENOME if there is an interval of Genome of length L in which this k-mer appears at least T times.) in a string\n",
    "\n",
    "def FindClumps(GENOME, K, L, T):\n",
    "    kmer_dict = {}\n",
    "    count = 0\n",
    "    for i in range(len(GENOME) - K + 1):\n",
    "        pattern = GENOME[i : i + K]\n",
    "        if pattern in kmer_dict:\n",
    "            kmer_dict[pattern].append(i)\n",
    "        else:\n",
    "            kmer_dict[pattern] = [i]\n",
    "    print(kmer_dict)\n",
    "    \n",
    "    for kmer, indices in kmer_dict.items():\n",
    "        if len(indices) >= T:\n",
    "            for j in range(len(indices) - T + 1):\n",
    "                total = indices[j + 2] - indices[j] + K\n",
    "                if total <= L:\n",
    "                    count = count + 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8812ae02-1291-40c9-a297-cba5fa1c2c06",
   "metadata": {},
   "source": [
    "#### Week 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7ad75b6-3e77-42e7-861e-83637609850f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function MinimumSkew(text) plots the diagram of the difference between the total number of occurrences\n",
    "# of G and the total number of occurrences of C in the genome genome and finds positions of the minimum values.\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def MinimumSkew(text):\n",
    "    text_spl = [x for x in text]\n",
    "    n = 0\n",
    "    skew_list = [0]\n",
    "    for i in text_spl:\n",
    "        if i == 'G':\n",
    "            n += 1\n",
    "            skew_list.append(n)\n",
    "        elif i == 'C':\n",
    "            n = n-1\n",
    "            skew_list.append(n)\n",
    "        else:\n",
    "            skew_list.append(n)\n",
    "    plt.plot(skew_list)\n",
    "    plt.title('The skew diagram')\n",
    "    plt.xlabel('Nucleotid position')\n",
    "    plt.ylabel('Skew')\n",
    "    plt.show()\n",
    "    list_min_cord = []\n",
    "    for j in range(len(skew_list)):\n",
    "        if skew_list[j] == min(skew_list):\n",
    "            list_min_cord.append(j)\n",
    "    return(list_min_cord)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af8fe33c-fef2-4be6-9eae-e95fe766a8b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function HammingDistance(text_1, text_2) calculates the Hamming distance (the number of mismatches\n",
    "# between strings) between two strings of the same length.\n",
    "\n",
    "def HammingDistance(text_1, text_2):\n",
    "    text_spl_1 = [x for x in text_1]\n",
    "    text_spl_2 = [x for x in text_2]\n",
    "    hamming_dist = 0\n",
    "    for i in range(len(text_1)):\n",
    "        if text_spl_1[i] != text_spl_2[i]:\n",
    "            hamming_dist += 1\n",
    "    return(hamming_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c38eea2-2c0d-4194-bb0a-a2e4a4998aa0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Function ApproximatePatternMatching(Pattern, Text, k) finds all approximate (with the HammingDistance <= k)\n",
    "# occurrences of a pattern in a string.\n",
    "\n",
    "def ApproximatePatternMatching(Pattern, Text, k):\n",
    "    ind_few_mismatches = []\n",
    "    text_list = []\n",
    "    for i in range(len(Text)-len(Pattern)+1):\n",
    "        text_list.append(Text[i:(i+len(Pattern))])\n",
    "        for j in text_list:\n",
    "            Spl_1 = [x for x in j]\n",
    "            Spl_Pattern = [x for x in Pattern]\n",
    "            hamming_dist = HammingDistance(Spl_1, Spl_Pattern)\n",
    "        if hamming_dist <= k:\n",
    "            ind_few_mismatches.append(i)\n",
    "    return(ind_few_mismatches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "932c079f-b5c6-45eb-a50d-492f09f640d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function ApproximatePatternCount(Pattern, Text, d) computes the Hamming distance between Pattern and\n",
    "# every k-mer substring of Text\n",
    "\n",
    "def ApproximatePatternCount(Pattern, Text, d):\n",
    "    count = 0\n",
    "    text_list = []\n",
    "    for i in range(len(Text)-len(Pattern)+1):\n",
    "        text_list.append(Text[i:(i+len(Pattern))])\n",
    "        for j in text_list:\n",
    "            Spl_1 = [x for x in j]\n",
    "            Spl_Pattern = [x for x in Pattern]\n",
    "            hamming_dist = HammingDistance(Spl_1, Spl_Pattern)\n",
    "        if hamming_dist <= d:\n",
    "            count += 1\n",
    "    return(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "330701bf-1651-49a7-9566-2975e9fbbdec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function Neighbors(Pattern, d) generates the d-neighborhood that is the set of all k-mers whose Hamming distance\n",
    "#from Pattern does not exceed d\n",
    "\n",
    "def Neighbors(Pattern, d):\n",
    "    nucleotides = ['A','C','G','T']\n",
    "    if d == 0:\n",
    "        return [Pattern]\n",
    "    if len(Pattern) == 1:\n",
    "        return nucleotides\n",
    "    Neighborhood = []\n",
    "    for neighbor in Neighbors(Pattern[1:],d):\n",
    "        if HammingDistance(Pattern[1:],neighbor) < d:\n",
    "            for nucl in nucleotides:\n",
    "                Neighborhood.extend([nucl+neighbor])\n",
    "        else:\n",
    "            Neighborhood.append(Pattern[0] + neighbor)\n",
    "    return Neighborhood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d53eed-066f-4082-8e65-8ac99d83fd10",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function FrequentWordsWithMismatches(Text, k, d) finds the most frequent k-mers with mismatches in a string\n",
    "\n",
    "def FrequentWordsWithMismatches(Text, k, d):\n",
    "    Patterns = []\n",
    "    freqMap = {}\n",
    "    n = len(Text)\n",
    "    for i in range(n-k+1):\n",
    "        Pattern = Text[i:(k+i)]\n",
    "        neighborhood = Neighbors(Pattern, d)\n",
    "        for j in range(len(neighborhood)-1):\n",
    "            neighbor = neighborhood[j]\n",
    "            if neighbor not in freqMap.keys():\n",
    "                freqMap[neighbor] = 1\n",
    "            else:\n",
    "                freqMap[neighbor] = freqMap[neighbor] + 1\n",
    "    m = max(freqMap.values())\n",
    "    for key_pat in freqMap:\n",
    "        if freqMap[key_pat] == m:\n",
    "            Patterns.append(key_pat)\n",
    "    return(Patterns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb7b4373-ab07-4d8b-8fa2-569642bb47f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function FreqWordsMismatch_and_ReverseCompl(Text, k, d) finds the most frequent k-mers (with mismatches and reverse complements) in a string\n",
    "\n",
    "def FreqWordsMismatch_and_ReverseCompl(Text, k, d):\n",
    "    Patterns = []\n",
    "    freqMap = {}\n",
    "    ReverseMap = {}\n",
    "    SumMap = {}\n",
    "    n = len(Text)\n",
    "    for i in range(n-k+1):\n",
    "        Pattern = Text[i:(k+i)]\n",
    "        neighborhood = Neighbors(Pattern, d)\n",
    "        for j in range(len(neighborhood)):\n",
    "            neighbor = neighborhood[j]\n",
    "            if neighbor not in freqMap.keys():\n",
    "                freqMap[neighbor] = 1\n",
    "            else:\n",
    "                freqMap[neighbor] = freqMap[neighbor] + 1\n",
    "        for l in range(len(neighborhood)):\n",
    "            neighbor_1 = ReverseComplement(neighborhood[l])\n",
    "            if neighbor_1 not in ReverseMap.keys():\n",
    "                ReverseMap[neighbor_1] = 1\n",
    "            else:\n",
    "                ReverseMap[neighbor_1] = ReverseMap[neighbor_1] + 1\n",
    "\n",
    "    for pat in freqMap.keys():\n",
    "        if pat in ReverseMap.keys():\n",
    "            SumMap[pat] = freqMap[pat] + ReverseMap[pat]\n",
    "    m_sum = max(SumMap.values())\n",
    "    for key_pat in SumMap.keys():\n",
    "        if SumMap[key_pat] == m_sum:\n",
    "            Patterns.append(key_pat)\n",
    "    return Patterns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9458a3f8-bd0c-478e-9089-d50c44c3431a",
   "metadata": {},
   "source": [
    "#### Week 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c3c9c54-b90c-41e9-9bc1-704a2d44cb46",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function MotifEnumeration(Dna, k, d) finds all (k, d)-motifs in a collection of strings.\n",
    "\n",
    "def MotifEnumeration(Dna, k, d):\n",
    "    Patterns = []\n",
    "    neighborhood = {}\n",
    "    all_kmers = {}\n",
    "    Dna_list = Dna.split(' ')\n",
    "    for str_num in range(len(Dna_list)):\n",
    "        all_kmers[str_num] = []\n",
    "        for i in range(len(Dna_list[str_num])-k+1):\n",
    "            all_kmers[str_num].append(Dna_list[str_num][i:i+k])\n",
    "\n",
    "        for num in list(all_kmers.keys()):\n",
    "            neighborhood[num] = []\n",
    "            for k_mer in all_kmers[num]:\n",
    "                neighborhood[num].extend(Neighbors(k_mer, d))\n",
    "                neighborhood[num] = list(set(neighborhood[num]))\n",
    "\n",
    "    count = {}\n",
    "    for k in neighborhood.keys():\n",
    "        for p in neighborhood[k]:\n",
    "            count[p] = 0\n",
    "    for key in neighborhood.keys():\n",
    "        for Pattern_mism in count.keys():\n",
    "            if Pattern_mism in neighborhood[key]:\n",
    "                count[Pattern_mism] += 1\n",
    "    \n",
    "    for k_mism in count.keys():\n",
    "        if count[k_mism] == len(Dna_list):\n",
    "            Patterns.append(k_mism)\n",
    "            \n",
    "    return Patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15df44af-a25d-4edf-8334-b044614fceee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function all_kmers_generator(Dna, k) generates all possible k-mers of k-length nucleotides.\n",
    "\n",
    "def all_kmers_generator(Dna, k):\n",
    "    Dna_list = Dna.split(' ')\n",
    "    all_kmers = {}\n",
    "    for str_num in range(len(Dna_list)):\n",
    "        all_kmers[Dna_list[str_num]] = []\n",
    "        for i in range(len(Dna_list[str_num])-k+1):\n",
    "            all_kmers[Dna_list[str_num]].append(Dna_list[str_num][i:i+k])\n",
    "    return all_kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "542619a1-b27c-4b71-ae66-a5de28151ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function DistanceBetweenPatternAndStrings(Pattern, Dna) finds a sum of distances between Pattern and each string in Dna\n",
    "\n",
    "def DistanceBetweenPatternAndStrings(Pattern, Dna):\n",
    "    k = len(Pattern)\n",
    "    distance = 0\n",
    "    all_kmers = all_kmers_generator(Dna, k)   \n",
    "\n",
    "    for str_pat in all_kmers.keys():\n",
    "        HammingDist = float('inf')\n",
    "        for k_mer_pat in all_kmers[str_pat]:\n",
    "            if HammingDist > HammingDistance(Pattern, k_mer_pat):\n",
    "                HammingDist = HammingDistance(Pattern, k_mer_pat)\n",
    "        distance = distance + HammingDist\n",
    "    return distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46567350-cdf0-40f5-aee7-d65c85aba5f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function MedianString(Dna, k) finds a median string (a k-mer Pattern that minimizes\n",
    "# DistanceBetweenPatternAndStrings(Pattern, Dna) among all possible choices of k-mers)\n",
    "\n",
    "import itertools\n",
    "from itertools import product\n",
    "\n",
    "def MedianString(Dna, k):\n",
    "    distance = float('inf')\n",
    "\n",
    "    perm_list = list(product(('A', 'T', 'C', 'G'), repeat=k))\n",
    "    perm_list_1 = ['']*len(perm_list)\n",
    "    for i in range(len(perm_list)):\n",
    "        perm_list[i] = list(perm_list[i])\n",
    "        for j in range(len(list(perm_list[i]))):\n",
    "            perm_list_1[i] = perm_list_1[i] + perm_list[i][j]\n",
    "    Patterns = perm_list_1\n",
    "\n",
    "    for t in range(len(Patterns)):\n",
    "        Pattern = Patterns[t]\n",
    "        if distance > DistanceBetweenPatternAndStrings(Pattern, Dna):\n",
    "            distance = DistanceBetweenPatternAndStrings(Pattern, Dna)\n",
    "            Median = Pattern\n",
    "    return Median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c72db9b2-12e6-4e21-90b3-790c1c0758d4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#To open the dataset\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "f = open(\"dataset_159_3.txt\", \"r+\", encoding='utf-8-sig')\n",
    "file = f.read()\n",
    "Text = file.split('\\n')[0]\n",
    "k = int(file.split('\\n')[1])\n",
    "Profile_0 = []\n",
    "Profile_1 = []\n",
    "for i in file.split('\\n')[2:]:\n",
    "    Profile_0.append(i)\n",
    "for j in range(len(Profile_0)):\n",
    "    Profile_1.append(Profile_0[j].split(' '))\n",
    "for l in range(len(Profile_1[0:-1])):\n",
    "    for h in range(len(Profile_1[l])):\n",
    "        Profile_1[l][h] = float(Profile_1[l][h])\n",
    "Profile_3 = Profile_1[0:-1]\n",
    "Nucl = ['A', 'C', 'G', 'T']\n",
    "Profile = dict(zip(Nucl, Profile_3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19a54148-bdbc-4a39-b872-5ec3bab7ab36",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function Profile_most_Probable_k_mer(Text, k, Profile) finds a Profile-most probable k-mer in a string.\n",
    "\n",
    "def Profile_most_Probable_k_mer(Text, k, Profile):\n",
    "    k_mers_list = list(all_kmers_generator(Text, k).values())[0]\n",
    "    sum_dict = {}\n",
    "    for k_mer in k_mers_list:\n",
    "        s = 0\n",
    "        for nucl_ind in range(len(k_mer)):\n",
    "            s = s + Profile[k_mer[nucl_ind]][nucl_ind]\n",
    "            sum_dict[k_mer] = s\n",
    "\n",
    "    most_prob_kmer = [k for k, v in sum_dict.items() if v == max(sum_dict.values())]\n",
    "    return most_prob_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1d93839-f18c-400b-b8c7-443c64770332",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function ProfileMostProbable_for_Greedy_Search(Text, k, Profile) is using for GreedyMotifSearch.\n",
    "\n",
    "def NumberToSymbol(index):\n",
    "    if index == 0:\n",
    "        return str(\"A\")\n",
    "    elif index == 1:\n",
    "        return str(\"C\")\n",
    "    elif index == 2:\n",
    "        return str(\"G\")\n",
    "    elif index == 3:\n",
    "        return str(\"T\")\n",
    "\n",
    "def window(s, k):\n",
    "    for i in range(1 + len(s) - k):\n",
    "        yield s[i:i+k]\n",
    "\n",
    "def SymbolToNumber(Symbol):\n",
    "    if Symbol == \"A\":\n",
    "        return 0\n",
    "    elif Symbol == \"C\":\n",
    "        return 1\n",
    "    elif Symbol == \"G\":\n",
    "        return 2\n",
    "    elif Symbol == \"T\":\n",
    "        return 3\n",
    "\n",
    "\n",
    "def ProfileMostProbable_for_Greedy_Search(Text, k, Profile):\n",
    "    letter = [[] for key in range(k)]\n",
    "    probable = \"\"\n",
    "    hamdict = {}\n",
    "    index = 1\n",
    "    for a in range(k):\n",
    "        for j in \"ACGT\":\n",
    "            letter[a].append(Profile[j][a])\n",
    "    for b in range(len(letter)):\n",
    "        number = max(letter[b])\n",
    "        probable += str(NumberToSymbol(letter[b].index(number)))\n",
    "    for c in window(Text, k):\n",
    "        for x in range(len(c)):\n",
    "            y = SymbolToNumber(c[x])\n",
    "            index *= float(letter[x][y])\n",
    "        hamdict[c] = index\n",
    "        index = 1\n",
    "    for pat, ham in hamdict.items():\n",
    "        if ham == max(hamdict.values()):\n",
    "            final = pat\n",
    "            break\n",
    "            \n",
    "    return final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e412315a-9202-455a-ab35-9e6ba820bc45",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function Score(Motif_list) finds score of motifs.\n",
    "\n",
    "import numpy as np\n",
    "from typing import List\n",
    "\n",
    "nuc_order = {'A': 0, 'C': 1,'G': 2,'T': 3}\n",
    "def Profile(Motif_list: List[str], k: int) -> np.ndarray:\n",
    "    Profile_1_dict = {'A': 0, 'C': 0,'G': 0,'T': 0}\n",
    "    prob_mat = np.zeros((4, k))\n",
    "    for nuc_seq in Motif_list:\n",
    "        prob_mat += fill_prob_mat(nuc_seq, k)\n",
    "    Profile_nump = prob_mat / np.sum(prob_mat, axis=0)\n",
    "    for key_num in range(len(Profile_1_dict.keys())):\n",
    "        Profile_1_dict[list(Profile_1_dict.keys())[key_num]] = list(Profile_nump[key_num, :])\n",
    "    return Profile_1_dict\n",
    "\n",
    "def fill_prob_mat(nuc_seq: str, k: int, nuc_order: dict = nuc_order) -> np.ndarray:\n",
    "    prob_mat = np.zeros((4, k))\n",
    "    for i, nuc in enumerate(nuc_seq):\n",
    "        prob_mat[nuc_order[nuc], i] += 1\n",
    "    return prob_mat\n",
    "\n",
    "def Score(Motif_list):\n",
    "    Sum = 0\n",
    "    for nuc_ind in range(len(Motif_list[0])):\n",
    "        nuc_col_dict = {}\n",
    "        for seq in Motif_list:\n",
    "            if seq[nuc_ind] not in nuc_col_dict.keys():\n",
    "                nuc_col_dict[seq[nuc_ind]] = 1\n",
    "            else:\n",
    "                nuc_col_dict[seq[nuc_ind]] += 1\n",
    "        count_col = len(Motif_list) - max(list(nuc_col_dict.values()))\n",
    "        Sum += count_col\n",
    "    return Sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ae8278-2ddf-4b9f-8fdf-892655c41bd9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Function GreedyMotifSearch(Dna, k, t) tests to see whether Motifs outscores the current best scoring collection of motifs.\n",
    "\n",
    "def GreedyMotifSearch(Dna, k, t):\n",
    "    BestMotifs = []\n",
    "    for string in Dna:\n",
    "        BestMotifs.append(string[0:k])\n",
    "\n",
    "    for nuc_i in range(len(Dna[0])-k+1):\n",
    "        Motif_list = []\n",
    "        Motif_list.append(Dna[0][nuc_i:nuc_i+k])\n",
    "\n",
    "        for dna_i in range(1,t):\n",
    "            Profile_1_dict = Profile(Motif_list[0:dna_i], k)\n",
    "            Motif_list.append(ProfileMostProbable_for_Greedy_Search(Dna[dna_i], k, Profile_1_dict))\n",
    "            \n",
    "        if Score(Motif_list) < Score(BestMotifs):\n",
    "            BestMotifs = Motif_list\n",
    "            \n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb123d1f-8020-4662-baa8-12eb6f2f49b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Greedy motif search with usage of pseudocounts (substitution zeroes with small numbers).\n",
    "\n",
    "import numpy as np\n",
    "from typing import List\n",
    "\n",
    "nuc_order = {'A': 0, 'C': 1,'G': 2,'T': 3}\n",
    "def Profile_with_pseudocounts(Motif_list: List[str], k: int) -> np.ndarray:\n",
    "    Profile_1_dict = {'A': 0, 'C': 0,'G': 0,'T': 0}\n",
    "    prob_mat = np.zeros((4, k))\n",
    "    for nuc_seq in Motif_list:\n",
    "        prob_mat += fill_prob_mat_with_pseudocounts(nuc_seq, k)\n",
    "    Profile_nump = prob_mat / np.sum(prob_mat, axis=0)\n",
    "    for key_num in range(len(Profile_1_dict.keys())):\n",
    "        Profile_1_dict[list(Profile_1_dict.keys())[key_num]] = list(Profile_nump[key_num, :])\n",
    "    return Profile_1_dict\n",
    "\n",
    "def fill_prob_mat_with_pseudocounts(nuc_seq: str, k: int, nuc_order: dict = nuc_order) -> np.ndarray:\n",
    "    prob_mat = np.zeros((4, k))\n",
    "    for j in range(4):\n",
    "        for g in range(k):\n",
    "            prob_mat[j,g] = 1\n",
    "    for i, nuc in enumerate(nuc_seq):\n",
    "        prob_mat[nuc_order[nuc], i] += 1\n",
    "    return prob_mat\n",
    "\n",
    "def GreedyMotifSearch_with_pseudocounts(Dna, k, t):\n",
    "    BestMotifs = []\n",
    "    for string in Dna:\n",
    "        BestMotifs.append(string[0:k])\n",
    "\n",
    "    for nuc_i in range(len(Dna[0])-k+1):\n",
    "        Motif_list = []\n",
    "        Motif_list.append(Dna[0][nuc_i:nuc_i+k])\n",
    "\n",
    "        for dna_i in range(1,t):\n",
    "            Profile_1_dict = Profile_with_pseudocounts(Motif_list[0:dna_i], k)\n",
    "            Motif_list.append(ProfileMostProbable_for_Greedy_Search(Dna[dna_i], k, Profile_1_dict))\n",
    "            \n",
    "        if Score(Motif_list) < Score(BestMotifs):\n",
    "            BestMotifs = Motif_list\n",
    "            \n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a814bc8-f15b-4806-bfde-01f343bee686",
   "metadata": {},
   "source": [
    "#### Week 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9df05651-310c-4222-b8f9-e537b19aae39",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function Random_k_mers_generation(Dna, k, t) creates a list of the randomly generated sequences from Dna.\n",
    "\n",
    "import random\n",
    "\n",
    "def Random_k_mers_generation(Dna, k, t):\n",
    "    Rand_k_mers = []\n",
    "    for i in range(len(Dna)):\n",
    "        rand_posit = random.randrange(len(Dna[0])-k)\n",
    "        Rand_k_mers.append(Dna[i][rand_posit:rand_posit+k])\n",
    "    return Rand_k_mers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89e63b99-c0d1-478d-a337-5bb32fcdc710",
   "metadata": {},
   "outputs": [],
   "source": [
    "def MostProbable_motifs(Profile, Dna, t, k):\n",
    "    MostProbable_motifs = []\n",
    "    \n",
    "    for dna_i in range(t):\n",
    "        MostProbable_motifs.append(ProfileMostProbable_for_Greedy_Search(Dna[dna_i], k, Profile))\n",
    "\n",
    "    return MostProbable_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ebda4f-b788-43de-a49b-038cfb762263",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function Randomized_GreedyMotifSearch(Dna, k, t) randomly selects possible locations and finds\n",
    "#a way to greedily change those locations until we have converged to the hidden motif.\n",
    "\n",
    "def Randomized_GreedyMotifSearch(Dna, k, t):\n",
    "    Motif_list = Random_k_mers_generation(Dna, k, t)\n",
    "    BestMotifs = Motif_list\n",
    "    while BestMotifs == Motif_list:\n",
    "        Profile_1_dict = Profile_with_pseudocounts(Motif_list, k)\n",
    "\n",
    "        Motif_list =  MostProbable_motifs(Profile_1_dict, Dna, t, k)\n",
    "            \n",
    "        if Score(Motif_list) < Score(BestMotifs):\n",
    "            BestMotifs = Motif_list\n",
    "            \n",
    "        else:\n",
    "            return BestMotifs\n",
    "\n",
    "#To find \n",
    "n = 1000\n",
    "Random_motifs = Randomized_GreedyMotifSearch(Dna, k, t)\n",
    "RandBestMotifs = Random_motifs\n",
    "\n",
    "for i in range(n+1):\n",
    "    Random_motifs = Randomized_GreedyMotifSearch(Dna, k, t)\n",
    "    if Score(Random_motifs) < Score(RandBestMotifs):\n",
    "         RandBestMotifs = Random_motifs\n",
    "    else:\n",
    "        BestMotifs_new = RandBestMotifs\n",
    "        \n",
    "BestMotifs_new #end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b30f5745-cb53-4f57-a9ea-f93baea15972",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cytometry_elizav_kernel",
   "language": "python",
   "name": "cytometry_elizav_kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
